---
title: "EDA dataset"
author: "Cheloshkina K"
date: "14 09 2019"
output:
  html_document:
    df_print: paged
    fig_caption: yes
    fig_height: 8
    fig_width: 10
  html_notebook: default
  pdf_document: default
---

```{r setup, include=FALSE, echo=FALSE}
# tinytex::install_tinytex()
library(dplyr)
library(reshape2)

library(ggplot2)
library(GGally)
library(wesanderson)
library(ggsci)
pal <- wes_palette("Zissou1", 100, type = "continuous")

win_len <- 100000

# load data
data <- read.csv(
  paste0("../data/datasets/dataset_", format(win_len, scientific = FALSE), ".csv")
  )

```


```{r}
sprintf("This markdown shows exploratory analysis of dataset for window length of %s bases.", format(win_len, big.mark = " ", scientific = FALSE))
```

## Hotspots statistics

Number of labeling strategies for each cancer type: 

```{r}

hsp_cols <- grep(x = names(data), pattern = "hsp",value = TRUE)
hsp_stats <- data.frame(hsp_name = hsp_cols, stringsAsFactors = FALSE)
hsp_stats$level <- unlist(lapply(strsplit(x = hsp_stats$hsp_name, split = "_"),
                                 function(x) x[2]))
hsp_stats$cancer_type <- unlist(lapply(strsplit(x = hsp_stats$hsp_name, split = "_"),
                                       function(x) x[3]))

hsp_stats %>%
  group_by(cancer_type) %>%
  summarize(n_hotspot_types = n())
```
Some cancer types have less hotspots types than others because some "neighbouring" labeling strategies (for example, 99% and 99.5%) result in the same labels. In such cases duplicates are eliminated.


Number of hotspots for each cancer type and labeling strategy.

```{r}
n_hotspots <- data %>%
  summarize_at(hsp_cols, sum) %>%
  t() %>%
  data.frame() 

names(n_hotspots) <- "n_hotspots"
n_hotspots$hsp_name <- rownames(n_hotspots)
rownames(n_hotspots)<- NULL

n_hotspots <- n_hotspots %>%
  inner_join(hsp_stats, by = "hsp_name")

dcast(data = n_hotspots, formula = cancer_type ~ level, value.var = "n_hotspots")
```

Here is a distribution of "all" hotspots among chromosomes for different cancer types. 

```{r}
hsp_all <- grep(x = hsp_cols, pattern = "all", value = TRUE)

chr_stats <- data %>% 
  group_by(chr) %>%
  summarize_at(hsp_all, sum)

total_hsp <- chr_stats %>%
  summarize_at(hsp_all, sum)

for (col in hsp_all){
  chr_stats[col] <- chr_stats[col] / as.numeric(total_hsp[col][1])
}

melted_chr_stats <- melt(chr_stats, id.vars = "chr")
melted_chr_stats$variable <- gsub(x = melted_chr_stats$variable, 
                                  pattern = "hsp_all_", replacement = "")
names(melted_chr_stats) <- c("chr", "cancer_type", "ratio_of_hotspots")

ggplot(melted_chr_stats, aes(cancer_type, chr)) +
  geom_tile(aes(fill = ratio_of_hotspots)) + scale_fill_gradientn(colours = pal)
```


Let's discover if there are cancer types which have the same hotspots for given ratio of windows marked as hotspots.

```{r}
dupl_cols <- duplicated(t(data[hsp_cols]))

if (any(dupl_cols)){
  dupl_cols <- hsp_cols[dupl_cols]
  duplicates <- list()
  for (dupl_col in dupl_cols){
    for (col in setdiff(hsp_cols, c(dupl_cols))){
      ifelse(all(data[col] == data[dupl_col]), 
             duplicates[dupl_col] <- col, FALSE)
    }
  }
  print(duplicates)  
}

```

## Features statistics

Conserved features (do not depend on tissue type):

```{r}

feat_names <- setdiff(names(data), c("chr","from","to", hsp_cols))
tissue_spec_feats <- grep(x = feat_names, pattern = "cancer", value = TRUE)
conserved_feats <- setdiff(feat_names, tissue_spec_feats)
print(length(feat_names))
conserved_feats
```

Tissue-specific features:

```{r}
un_tissue_spec_feats <- unique(unlist(lapply(
  strsplit(x = tissue_spec_feats, split = "_"), 
  function(x) paste(x[3:length(x)], collapse = "_")
)))
un_tissue_spec_feats
```

```{r}
unique(hsp_stats$cancer_type)

l <- lapply(
  strsplit(x = tissue_spec_feats, split = "_"), 
  function(x) c(x[2], paste(x[3:length(x)], collapse = "_")
))

df <- data.frame(do.call("rbind", l))
names(df) <- c("cancer_type", "feature")
df$value <- "+"
dcast(df, formula = feature ~ cancer_type, value.var = "value")

```

As transcription factors are available only for liver, we will use them for prediction of all cancer types hotspots.

```{r}
new_conserved <- c(
  "A_Phased_Repeat", "Direct_Repeat", 
  "Inverted_Repeat", "Mirror_Repeat", "Short_Tandem_Repeat", 
  "Z_DNA_Motif", "G_quadruplex", 
  "stemloops_16_50", "stemloops_6_15", 
  "X3UTR", "X5UTR", "codingExons", "downstream", "introns", "promoters", "WholeGenes", 
  "tad_boundaries_liver", "tad_boundaries_ovary", "tad_boundaries_pancreatic",
  
  "cancer_liver_ATF3.human", "cancer_liver_CTCF.human", "cancer_liver_EGR1.human",
  "cancer_liver_FOXA1.human", "cancer_liver_FOXA2.human", "cancer_liver_GABPA.human",
  "cancer_liver_HNF4A.human", "cancer_liver_HNF4G.human", "cancer_liver_JUND.human",
  "cancer_liver_MAX.human", "cancer_liver_NR2F2.human", "cancer_liver_REST.human",
  "cancer_liver_RXRA.human", "cancer_liver_SP1.human", "cancer_liver_YY1.human",
  "cancer_liver_ZBTB33.human"
)

new_tissue_spec <- setdiff(feat_names, new_conserved)
```

Correlations between features:

```{r fig.width=15, fig.height=10}

cormat <- round(cor(data[feat_names], method = "spearman"), 2)

reorder_cormat <- function(cormat){
  dd <- as.dist((1-cormat)/2)
  hc <- hclust(dd)
  cormat <- cormat[hc$order, hc$order]
}
cormat <- reorder_cormat(cormat)
melted_cormat <- melt(cormat, na.rm = TRUE)

ggplot(melted_cormat, aes(Var2, Var1, fill = value))+
  geom_tile()+
  theme_minimal()+ # minimal theme
  theme(axis.text.x = element_text(angle = 45, vjust = 1, 
    size = 12, hjust = 1))  + scale_fill_gradientn(colours = pal)
```

```{r include=FALSE, echo=FALSE}
pancr_ind <- grep(melted_cormat$Var1, pattern = "pancreas")
melted_cormat <- melted_cormat[-pancr_ind,]
pancr_ind <- grep(melted_cormat$Var2, pattern = "pancreas")
melted_cormat <- melted_cormat[-pancr_ind,]

g1 <- ggplot(melted_cormat, aes(Var2, Var1, fill = value))+
  geom_tile() +
  xlab("Feature") + 
  ylab("Feature") +
  theme_minimal()+ # minimal theme
  theme(
    text = element_text(size = 8, family = "arial"),
    axis.title.x = element_text(size = 10),
    axis.title.y = element_text(size = 10),
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
    axis.text.y = element_text(hjust = 1)
    ) + 
  labs(fill = "Spearman \ncorrelation")+
  scale_fill_gradientn(colours = pal) + 
  coord_fixed()

# ggsave(filename = "Fig1.tiff", plot = g1, device = "tiff", path = "../paper_materials/supplementary figures/",
#        dpi = 600, units = "in", width = 6, height = 5, scale = 2, compression = "lzw")

```


Distribution of correlation value of all pairs of features

```{r}
ggplot(melted_cormat, aes(value)) + 
  geom_histogram(bins = 50)

```

Distribution of mean correlation between groups of features

```{r}
feat_groups <- data.frame(
  feat = c("A_Phased_Repeat","Direct_Repeat", "Inverted_Repeat", "Mirror_Repeat","Short_Tandem_Repeat",
             "G_quadruplex", "stemloops_16_50", "stemloops_6_15", "Z_DNA_Motif"),
  feat_group = "sec_str" 
)
feat_groups <- rbind(feat_groups, data.frame(
  feat = c("X3UTR", "X5UTR", "codingExons", "downstream", "introns", "promoters", "WholeGenes"),
  feat_group = "reg"   
))
feat_groups <- rbind(feat_groups, data.frame(
  feat = c("tad_boundaries_liver", "tad_boundaries_ovary", "tad_boundaries_pancreatic"),
  feat_group = "tad"   
))
feat_groups <- rbind(feat_groups, data.frame(
  feat = c("cancer_skin_DNase_seq", "cancer_brain_DNase_seq", "cancer_blood_DNase_seq",
              "cancer_prostate_DNase_seq", "cancer_pancreas_DNase_seq", "cancer_ovary_DNase_seq",
              "cancer_liver_DNase_seq", "cancer_breast_DNase_seq", "cancer_uterus_DNase_seq",
              "cancer_bone_DNase_seq"), 
  feat_group = "chromatin"   
)) 
feat_groups <- rbind(feat_groups, data.frame(
  feat = c("cancer_brain_DNA_methylation", "cancer_breast_DNA_methylation",
            "cancer_liver_DNA_methylation", "cancer_pancreas_DNA_methylation",
            "cancer_skin_DNA_methylation", "cancer_uterus_DNA_methylation"), 
  feat_group = "methyl"   
)) 
feat_groups <- rbind(feat_groups, data.frame(
  feat = c( "cancer_liver_H3K27ac.human", "cancer_uterus_H3K27ac.human", "cancer_blood_H3K27ac.human",
               "cancer_brain_H3K27ac.human",
               "cancer_pancreas_H3K27ac.human","cancer_blood_H3K27me3.human","cancer_brain_H3K27me3.human",    
               "cancer_blood_H3K36me3.human","cancer_pancreas_H3K36me3.human","cancer_brain_H3K36me3.human",    
               "cancer_pancreas_H3K4me1.human","cancer_blood_H3K4me1.human","cancer_brain_H3K4me1.human",
               "cancer_breast_H3K4me3.human","cancer_uterus_H3K4me3.human","cancer_liver_H3K4me3.human",
               "cancer_brain_H3K4me3.human","cancer_blood_H3K4me3.human","cancer_skin_H3K4me3.human",
               "cancer_pancreas_H3K4me3.human","cancer_brain_H3K9me3.human","cancer_pancreas_H3K9me3.human",
               "cancer_blood_H3K9me3.human" ), 
  feat_group = "histones"   
)) 
feat_groups <- rbind(feat_groups, data.frame(
  feat = c("cancer_liver_ATF3.human", "cancer_liver_CTCF.human", "cancer_pancreas_CTCF.human",
        "cancer_liver_EGR1.human", "cancer_liver_FOXA1.human", "cancer_liver_FOXA2.human", 
        "cancer_liver_GABPA.human","cancer_liver_HNF4A.human", "cancer_liver_HNF4G.human",
        "cancer_liver_JUND.human","cancer_liver_MAX.human", "cancer_liver_NR2F2.human",
        "cancer_liver_REST.human", "cancer_liver_RXRA.human", "cancer_liver_SP1.human",
        "cancer_liver_YY1.human", "cancer_liver_ZBTB33.human"), 
  feat_group = "tf"   
))

melted_cormat <- melted_cormat %>%
  inner_join(feat_groups, by=c("Var1" = "feat")) %>%
  rename("feat_group1" = "feat_group") %>%
  inner_join(feat_groups, by=c("Var2" = "feat")) %>%
  rename("feat_group2" = "feat_group")
melted_cormat <- melted_cormat[melted_cormat$Var1 != melted_cormat$Var2, ]
ggplot(data=melted_cormat, aes(value))+
  geom_density()+
  facet_grid(feat_group1 ~ feat_group2)

```

```{r}
feat_group_stats <- melted_cormat %>%
  group_by(feat_group1, feat_group2) %>%
  summarize(
    q1 = quantile(value, 0.25),
    q2 = quantile(value, 0.5),
    q3 = quantile(value, 0.75)
  )

feat_group_stats %>%
  filter(feat_group1 == feat_group2) %>%
  arrange(q1)
```

```{r}
feat_group_stats %>%
  filter(feat_group1 != feat_group2) %>%
  arrange(feat_group1, q1)
```


```{r}
feat_group_stats %>%
  filter(feat_group1 != feat_group2) %>%
  group_by(feat_group1) %>%
  filter(q3 == max(q3)) %>%
  select(-c("q1", "q2")) %>%
  arrange(q3)
```
```{r}
feat_group_stats %>%
  filter(feat_group1 != feat_group2) %>%
  group_by(feat_group1) %>%
  filter(q1 == max(q1)) %>%
  select(-c("q3", "q2")) %>%
  arrange(q1)
```
```{r}
feat_group_stats %>%
  group_by(feat_group1,feat_group2) %>%
  summarize(med = median(q2)) %>%
  ggplot(aes(x=feat_group1, y=feat_group2, fill=med)) + geom_tile()  + scale_fill_gradientn(colours = pal)
```

Distribution of features

```{r}
feat_melted <- melt(data[feat_names])

sec_str <- c("A_Phased_Repeat", "Direct_Repeat", 
  "Inverted_Repeat", "Mirror_Repeat", "Short_Tandem_Repeat", 
  "Z_DNA_Motif", "G_quadruplex", 
  "stemloops_16_50", "stemloops_6_15")

ggplot(feat_melted[feat_melted$variable %in% sec_str, ], aes(value)) + 
  geom_histogram(bins=30) + 
  facet_wrap(~ variable, scales = "free", nrow = 3) 
```

```{r}
ggplot(feat_melted[feat_melted$variable %in% sec_str, ], aes(value)) + 
  geom_histogram(bins=30) + 
  facet_wrap(~ variable, scales = "free", nrow = 3) + 
  scale_x_continuous(trans = 'log1p')
```


``` {r}
reg_tad <- c(  
  "X3UTR", "X5UTR", "codingExons", "downstream", "introns", "promoters", "WholeGenes", 
  "tad_boundaries_liver", "tad_boundaries_ovary", "tad_boundaries_pancreatic")

ggplot(feat_melted[feat_melted$variable %in% reg_tad, ], aes(value)) + 
  geom_histogram(bins=30) + 
  facet_wrap(~ variable, scales = "free", nrow = 3) 
```



``` {r}
ggplot(feat_melted[feat_melted$variable %in% reg_tad, ], aes(value)) + 
  geom_histogram(bins=50) + 
  facet_wrap(~ variable, scales = "free", nrow = 3)  + 
  scale_x_continuous(trans = 'log1p')
```
  
``` {r}
tf <- c(  
  "cancer_liver_ATF3.human", "cancer_liver_CTCF.human", "cancer_liver_EGR1.human",
  "cancer_liver_FOXA1.human", "cancer_liver_FOXA2.human", "cancer_liver_GABPA.human",
  "cancer_liver_HNF4A.human", "cancer_liver_HNF4G.human", "cancer_liver_JUND.human",
  "cancer_liver_MAX.human", "cancer_liver_NR2F2.human", "cancer_liver_REST.human",
  "cancer_liver_RXRA.human", "cancer_liver_SP1.human", "cancer_liver_YY1.human",
  "cancer_liver_ZBTB33.human")

ggplot(feat_melted[feat_melted$variable %in% tf, ], aes(value)) + 
  geom_histogram(bins=30) + 
  facet_wrap(~ variable, scales = "free", nrow = 4) 
```
  
``` {r}
ggplot(feat_melted[feat_melted$variable %in% tf, ], aes(value)) + 
  geom_histogram(bins=50) + 
  facet_wrap(~ variable, scales = "free", nrow = 4)  + 
  scale_x_continuous(trans = 'log1p')
```



``` {r}
dnase <- c(  
  "cancer_skin_DNase_seq", "cancer_brain_DNase_seq", "cancer_blood_DNase_seq", 
  "cancer_prostate_DNase_seq", "cancer_pancreas_DNase_seq", "cancer_ovary_DNase_seq", 
  "cancer_liver_DNase_seq", "cancer_breast_DNase_seq", "cancer_uterus_DNase_seq", 
  "cancer_bone_DNase_seq")

ggplot(feat_melted[feat_melted$variable %in% dnase, ], aes(value)) + 
  geom_histogram(bins=30) + 
  facet_wrap(~ variable, scales = "free", nrow = 4) 
```
  
``` {r}
ggplot(feat_melted[feat_melted$variable %in% dnase, ], aes(value)) + 
  geom_histogram(bins=50) + 
  facet_wrap(~ variable, scales = "free", nrow = 4)  + 
  scale_x_continuous(trans = 'log1p')
```


``` {r}
meth <- c(  
  "cancer_brain_DNA_methylation", "cancer_breast_DNA_methylation",
  "cancer_liver_DNA_methylation", "cancer_pancreas_DNA_methylation",
  "cancer_skin_DNA_methylation", "cancer_uterus_DNA_methylation")

ggplot(feat_melted[feat_melted$variable %in% meth, ], aes(value)) + 
  geom_histogram(bins=30) + 
  facet_wrap(~ variable, scales = "free", nrow = 4) 
```
  
``` {r}
ggplot(feat_melted[feat_melted$variable %in% meth, ], aes(value)) + 
  geom_histogram(bins=50) + 
  facet_wrap(~ variable, scales = "free", nrow = 4)  + 
  scale_x_continuous(trans = 'log1p')
```

``` {r}
h27 <- c(  
   "cancer_blood_H3K27ac.human","cancer_brain_H3K27ac.human",
   "cancer_pancreas_H3K27ac.human", "cancer_blood_H3K27me3.human", "cancer_brain_H3K27me3.human")

ggplot(feat_melted[feat_melted$variable %in% h27, ], aes(value)) + 
  geom_histogram(bins=30) + 
  facet_wrap(~ variable, scales = "free", nrow = 2) 
```
  
``` {r}
ggplot(feat_melted[feat_melted$variable %in% h27, ], aes(value)) + 
  geom_histogram(bins=50) + 
  facet_wrap(~ variable, scales = "free", nrow = 2)  + 
  scale_x_continuous(trans = 'log1p')
```



``` {r}
hk4 <- c(  
  "cancer_pancreas_H3K4me1.human", "cancer_blood_H3K4me1.human", "cancer_brain_H3K4me1.human",
  "cancer_breast_H3K4me3.human", "cancer_uterus_H3K4me3.human", "cancer_liver_H3K4me3.human",
  "cancer_brain_H3K4me3.human", "cancer_blood_H3K4me3.human", "cancer_skin_H3K4me3.human",
  "cancer_pancreas_H3K4me3.human"  )

ggplot(feat_melted[feat_melted$variable %in% hk4, ], aes(value)) + 
  geom_histogram(bins=30) + 
  facet_wrap(~ variable, scales = "free", nrow = 3) 
```
  
``` {r}
ggplot(feat_melted[feat_melted$variable %in% hk4, ], aes(value)) + 
  geom_histogram(bins=50) + 
  facet_wrap(~ variable, scales = "free", nrow = 3)  + 
  scale_x_continuous(trans = 'log1p')
```




``` {r}
hrest <- c(
  "cancer_blood_H3K36me3.human", "cancer_pancreas_H3K36me3.human", "cancer_brain_H3K36me3.human",
  "cancer_brain_H3K9me3.human", "cancer_pancreas_H3K9me3.human", "cancer_blood_H3K9me3.human")

ggplot(feat_melted[feat_melted$variable %in% hrest, ], aes(value)) + 
  geom_histogram(bins=30) + 
  facet_wrap(~ variable, scales = "free", nrow = 2) 
```
  
``` {r}
ggplot(feat_melted[feat_melted$variable %in% hrest, ], aes(value)) + 
  geom_histogram(bins=50) + 
  facet_wrap(~ variable, scales = "free", nrow = 2)  + 
  scale_x_continuous(trans = 'log1p')
```

For 1 Mb windows and 99.99 there is only 1 hotspot for each cancer type. Interestingly,it is the same window for blood, bone,breast, ovary and uterus cancer.

```{r}

# load 1Mb data
data <- read.csv("../data/datasets/dataset_1000000.csv")

hsp <- data[data$hsp_99.99._blood == 1, ]

hsp_cols <- grep(x = names(data), pattern = "hsp",value = TRUE)
feat_names <- setdiff(names(data), c(hsp_cols, "chr", "from", "to"))
hs_q <- data.frame()
for (f in feat_names){
  cf <- ecdf(as.vector(data[, f]))
  q <- cf(as.numeric(hsp[f]))
  hs_q <- rbind(hs_q, data.frame('feat' = f, 'q' = q))
}

sec_str <- c("A_Phased_Repeat","Direct_Repeat", "Inverted_Repeat", "Mirror_Repeat","Short_Tandem_Repeat",
             "G_quadruplex", "stemloops_16_50", "stemloops_6_15", "Z_DNA_Motif" )
reg <- c("X3UTR", "X5UTR", "codingExons", "downstream", "introns", "promoters", "WholeGenes")
tad <- c("tad_boundaries_liver", "tad_boundaries_ovary", "tad_boundaries_pancreatic")

chromatin <-c("cancer_skin_DNase_seq", "cancer_brain_DNase_seq", "cancer_blood_DNase_seq",
              "cancer_prostate_DNase_seq", "cancer_pancreas_DNase_seq", "cancer_ovary_DNase_seq",
              "cancer_liver_DNase_seq", "cancer_breast_DNase_seq", "cancer_uterus_DNase_seq",
              "cancer_bone_DNase_seq")
methyl <- c("cancer_brain_DNA_methylation", "cancer_breast_DNA_methylation",
            "cancer_liver_DNA_methylation", "cancer_pancreas_DNA_methylation",
            "cancer_skin_DNA_methylation", "cancer_uterus_DNA_methylation")
histones <- c( "cancer_liver_H3K27ac.human", "cancer_uterus_H3K27ac.human", "cancer_blood_H3K27ac.human",
               "cancer_brain_H3K27ac.human",
               "cancer_pancreas_H3K27ac.human","cancer_blood_H3K27me3.human","cancer_brain_H3K27me3.human",    
               "cancer_blood_H3K36me3.human","cancer_pancreas_H3K36me3.human","cancer_brain_H3K36me3.human",    
               "cancer_pancreas_H3K4me1.human","cancer_blood_H3K4me1.human","cancer_brain_H3K4me1.human",
               "cancer_breast_H3K4me3.human","cancer_uterus_H3K4me3.human","cancer_liver_H3K4me3.human",
               "cancer_brain_H3K4me3.human","cancer_blood_H3K4me3.human","cancer_skin_H3K4me3.human",
               "cancer_pancreas_H3K4me3.human","cancer_brain_H3K9me3.human","cancer_pancreas_H3K9me3.human",
               "cancer_blood_H3K9me3.human" )
tf <- c("cancer_liver_ATF3.human", "cancer_liver_CTCF.human", "cancer_pancreas_CTCF.human",
        "cancer_liver_EGR1.human", "cancer_liver_FOXA1.human", "cancer_liver_FOXA2.human", 
        "cancer_liver_GABPA.human","cancer_liver_HNF4A.human", "cancer_liver_HNF4G.human",
        "cancer_liver_JUND.human","cancer_liver_MAX.human", "cancer_liver_NR2F2.human",
        "cancer_liver_REST.human", "cancer_liver_RXRA.human", "cancer_liver_SP1.human",
        "cancer_liver_YY1.human", "cancer_liver_ZBTB33.human"  )

hs_q$group <- ''
hs_q[hs_q$feat %in% sec_str, 'group'] <- "secondary structure"
hs_q[hs_q$feat %in% reg, 'group'] <- "genome regions"
hs_q[hs_q$feat %in% tad, 'group'] <- "TAD"
hs_q[hs_q$feat %in% chromatin, 'group'] <- "chromatin"
hs_q[hs_q$feat %in% methyl, 'group'] <- "methylation"
hs_q[hs_q$feat %in% histones, 'group'] <- "histones modifications"
hs_q[hs_q$feat %in% tf, 'group'] <- "transcription factors"

hs_q %>%
  group_by(group) %>%
  summarize(mean_q = mean(q)) %>%
  arrange(mean_q)
```

```{r}
hs_q %>%
  group_by(group) %>%
  summarize(max_q = max(q)) %>%
  arrange(max_q)

```


```{r}
hs_q %>%
  group_by(group) %>%
  summarize(min_q = min(q)) %>%
  arrange(min_q)

```